Added UI selectors, actions & reducers for editing shipping methods inside a shipping zone.

DanReyLop opened the pull request:
Fixes #14273 

Please don't freak out about this PR size. Of those 1500 added lines, about 1000 of them are tests.

This PR adds UI-related reducers, actions and selectors for editing the shipping methods configured inside a shipping zone. This includes:
* Changing a shipping method type (which really just deletes it and creates a new one, preserving its order).
* Adding a new shipping method to the zone (it won't allow repeated shipping methods of the same type inside a zone, except for `Flat Rate`).
* Editing settings for the built-in shipping methods (Free Shipping, Local Pickup, Flat Rate).
* Removing a shipping method from a zone.

**To test**: This code isn't used anywhere, so just run the tests.

ryelle commented:
Can you add a `README` to explain what the state sub-tree(s) should look like, and what the different actions do? I'm having trouble following it through multiple files. See #14599 for some examples.

We got rid of the `payload` pattern in #14737, so all those actions can be flattened. Also, if you can, I'd suggest using a more descriptive name than `id` — in most of those actions I'm still not sure what it's the ID of.

DanReyLop commented:
> We got rid of the payload pattern in #14737, so all those actions can be flattened. Also, if you can, I'd suggest using a more descriptive name than id — in most of those actions I'm still not sure what it's the ID of.

Done and done.

> Can you add a README to explain what the state sub-tree(s) should look like, and what the different actions do? I'm having trouble following it through multiple files. See #14599 for some examples.

Also done :) Thanks for the examples! I missed that PR.
I've added JSDoc comments to the actions instead of putting them in a separate `README.md` file, I feel that if we put documantation outside of the code it will get outdated pretty fast.

ryelle commented on client/extensions/woocommerce/state/ui/shipping/zones/README.md:
```diff
@@ -0,0 +1,77 @@
+	zones: {
+		creates: [
+			{ // 1 new zone has been created:
+				id: { index: 0 }, // Temporary ID, since this zone still doesn't have a real ID provided by the API
```
This is possibly a simple thing I'm having trouble understanding — why is the temporary ID an object?

DanReyLop replied:
Basically to differentiate it from a "regular" ID. I guess we could have chosen something different like negative numbers or something like that. When calling `state/ui/helpers/nextBucketIndex()` with the `creates` collection, it will return a new unique ID object.

It took a while to sink in for me, I suggest looking at a simpler UI reducer like the `ui/products/edits-reducer.js`. Also, we should probably document better this whole deal.

ryelle commented on client/extensions/woocommerce/state/ui/shipping/zones/methods/local-pickup/actions.js:
```diff
@@ -0,0 +1,41 @@
+	return {
+		type: WOOCOMMERCE_SHIPPING_ZONE_METHOD_SET_TAXABLE,
+		siteId,
+		methodType: 'flat_rate',
```
Should this be `local_pickup`?

DanReyLop replied:
Yes, my bad, copy&paste. Fixed.

DanReyLop commented:
> If method_id was changed to methodType, could we also change method_title just so everything is camel case?

@marcinbot `method_title` shouldn't be used. Use `title` instead. `method_title` is the name of the method (`Flat Rate`, `Free Shipping`), and `title` is the merchant-defined name for it.

> Also, new methods are created with their own properties in the root object, for example {title, type, taxable,minSpend}, whereas the existing ones come from the server (and are passed to the UI) in this format: {title, type, settings:{taxable,minSpend}}. Can this be more consistent?

Since we aren't data-driving anything, it won't hurt to discard all the returned properties except for the settings values, give me a few minutes.

DanReyLop commented on client/extensions/woocommerce/state/ui/shipping/zones/methods/selectors.js:
```diff
@@ -0,0 +1,159 @@
+import { builtInShippingMethods } from 'woocommerce/state/ui/shipping/zones/methods/reducer';
+import { mergeMethodEdits } from './helpers';
+
+const getShippingZone = ( state, zoneId, siteId ) => {
```
Not really. This is a small internal helper function. It's not `export`ed.

DanReyLop commented:
> Also, new methods are created with their own properties in the root object, for example {title, type, taxable,minSpend}, whereas the existing ones come from the server (and are passed to the UI) in this format: {title, type, settings:{taxable,minSpend}}. Can this be more consistent?

@marcinbot See https://github.com/Automattic/wp-calypso/pull/14890/commits/02869cd8c2620156f4bbdff72e56a43ed4e0e4df

DanReyLop commented on client/extensions/woocommerce/state/ui/shipping/zones/methods/reducer.js:
```diff
@@ -0,0 +1,118 @@
+reducer[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_EDIT_TITLE ] = ( state, { methodId, title } ) => {
+	const bucket = getBucket( { id: methodId } );
+	const index = findIndex( state[ bucket ], { id: methodId } );
+	if ( -1 === index ) {
```
Fixed in https://github.com/Automattic/wp-calypso/pull/14890/commits/8f0da04521adca319e8ed9894a626b4cb5eec679

DanReyLop commented on client/extensions/woocommerce/state/ui/shipping/zones/methods/selectors.js:
```diff
@@ -0,0 +1,159 @@
+	if ( ! areShippingZonesLoaded( state, siteId ) ) {
+		return [];
+	}
+	const zone = getShippingZone( state, zoneId, siteId );
```
Fixed in https://github.com/Automattic/wp-calypso/pull/14890/commits/411893092b25286d1a6014c9b285f2c7253bd95d

DanReyLop commented:
> Looks like you also removed the enabled field from the parent object. The designs didn't account for it, but is it worth keeping it and adding appropriate actions for changing it? I've been modifying the UI with this field in mind. I feel it's pretty important.

You're 100% right. I was so fixed on the designs that I forgot that the API can return disabled methods. I'll add that.

DanReyLop commented on client/extensions/woocommerce/state/ui/shipping/zones/methods/flat-rate/reducer.js:
```diff
@@ -0,0 +1,29 @@
+
+reducer[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_SET_TAXABLE ] = ( state, { isTaxable } ) => {
+	return { ...state,
+		taxable: isTaxable ? 'Taxable' : 'None',
```
My bad, I always intended to match server-side names. Fixed in https://github.com/Automattic/wp-calypso/pull/14890/commits/5ac6a1e9ba46df54a67443bc7990f088c681a9ff

DanReyLop commented:
> Looks like you also removed the enabled field from the parent object. The designs didn't account for it, but is it worth keeping it and adding appropriate actions for changing it? I've been modifying the UI with this field in mind. I feel it's pretty important.

I've created https://github.com/Automattic/wp-calypso/pull/15135 to track this since this PR is already huge and I don't want to drag it any longer. I've just rebased this PR with `master` so you should do the same (or wait until this is released).

If (when) you run into more issues wiring the UI, I'll fix them in separate PRs.
